src/sdk/pynni/nni/bohb_advisor/config_generator.py

```diff
@@ -0,0 +1,326 @@
+            fraction of random configurations returned
+        bandwidth_factor: float
+            widens the bandwidth for contiuous parameters for proposed points to optimize EI
+        min_bandwidth: float
+            to keep diversity, even when all (good) samples have the same value for one of the parameters,
+            a minimum bandwidth (Default: 1e-3) is used instead of zero.
+
+        """
+        self.top_n_percent = top_n_percent
+        self.configspace = configspace
+        self.bw_factor = bandwidth_factor
+        self.min_bandwidth = min_bandwidth
+
+        self.min_points_in_model = min_points_in_model
+        if min_points_in_model is None:
+            self.min_points_in_model = len(self.configspace.get_hyperparameters())+1
+
+        if self.min_points_in_model < len(self.configspace.get_hyperparameters())+1:
+            logger.warning('Invalid min_points_in_model value. Setting it to %i'%(len(self.configspace.get_hyperparameters())+1))
+            self.min_points_in_model =len(self.configspace.get_hyperparameters())+1
+
+        self.num_samples = num_samples
+        self.random_fraction = random_fraction
+
+        hps = self.configspace.get_hyperparameters()
+
+        self.kde_vartypes = ""
+        self.vartypes = []
+
+        for h in hps:
+            if hasattr(h, 'choices'):
+                self.kde_vartypes += 'u'
+                self.vartypes += [len(h.choices)]
+            else:
+                self.kde_vartypes += 'c'
+                self.vartypes += [0]
+
+        self.vartypes = np.array(self.vartypes, dtype=int)
+
+        # store precomputed probs for the categorical parameters
+        self.cat_probs = []
+
+        self.configs = dict()
+        self.losses = dict()
+        self.good_config_rankings = dict()
+        self.kde_models = dict()
+
+    def largest_budget_with_model(self):
+        if len(self.kde_models) == 0:
+            return(-float('inf'))
+        return(max(self.kde_models.keys()))
+
+    def get_config(self, budget):
+        """Function to sample a new configuration
+        This function is called inside BOHB to query a new configuration
+
+        Parameters:
+        -----------
+        budget: float
+            the budget for which this configuration is scheduled
+
+        Returns
+        -------
+        config
+            should return a valid configuration
+        """
+        logger.debug('start sampling a new configuration.')
+        sample = None
+        info_dict = {}
+
+        # If no model is available, sample from prior
+        # also mix in a fraction of random configs
+        if len(self.kde_models.keys()) == 0 or np.random.rand() < self.random_fraction:
+            sample = self.configspace.sample_configuration()
+            info_dict['model_based_pick'] = False
+
+        best = np.inf
+        best_vector = None
+
+        if sample is None:
+            try:
+                # sample from largest budget
```

Comment:
I think this code is too long. Could you split it to few functions?

Reply:
Done